{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare QTLs for tomato fruit shape & potato tuber shape\n",
    "\n",
    "## Background\n",
    "\n",
    "Reference tomatoes have a round fruit shape while the reference potatoes have an elongated tuber shape. Recently published article (<a href=\"https://dx.doi.org/10.1038%2Fs41467-018-07216-8\">Wu et al., 2018</a>) reveals that the SLOFP20 gene present on the tomato chromosome 10 is responsible for round fruits. However, this gene does not have an ortholog in the reference potato genome (DM), which results in elongated tuber. This notebook is used to map the QTL regions in both tomato and potato. First, genes in the QTLs are classified into three categories: i) corresponding genes (orthologs) between tomato and potato, ii) tomato genes with no orthologs in potato and iii) vice versa. For each category, GO annotations are retrieved and then compared among the categories."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from diagrams import DIAGRAMS\n",
    "from IPython.display import Image,SVG\n",
    "import requests, pandas as pd\n",
    "\n",
    "url = \"http://pbg-ld.candygene-nlesc.surf-hosted.nl:8088/api/candYgene/queries\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "QTL for fruit shape in tomatoes on chromosome 10 is marked by the genes Solyc10g075170 and Solyc10g076240. Similarly QTL for tuber shape in potatoes on chromosome 10 is marked by the genes PGSC0003DMG400006678 and PGSC0003DMG400020801.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#potato\n",
    "pg1 = \"PGSC0003DMG400006678\"\n",
    "pg2 = \"PGSC0003DMG400020801\"\n",
    "#tomato\n",
    "tg1 = \"Solyc10g075170.1\"\n",
    "tg2 = \"Solyc10g076240.1\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tomato\n",
    "\n",
    "### Find co-ordinates of the QTL interval\n",
    "Search locations of genes, and compute interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chrom': 'chromosome 10',\n",
       " 'taxon_id': '4081',\n",
       " 'begin_pos': 58895882,\n",
       " 'end_pos': 59082990}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    resp1 = requests.get(url+\"/getFeatureLocation\", \n",
    "                        params={\"featureid\": \"'\"+tg1+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "    resp2 = requests.get(url+\"/getFeatureLocation\", \n",
    "                        params={\"featureid\": \"'\"+tg2+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "    intervalT = {}\n",
    "    intervalT[\"chrom\"] = resp1.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "    intervalT[\"taxon_id\"] = resp1.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "    intervalT[\"begin_pos\"] = int(resp1.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"])\n",
    "    intervalT[\"end_pos\"] = int(resp2.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])    \n",
    "except:\n",
    "    raise Exception(\"couldn't get interval for \"+str(tg1)+\" or \"+str(tg2))\n",
    "intervalT    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List the candidate genes inside this QTL interval and search potato orthologs\n",
    "List genes and potato orthologs inside this interval"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find tomato genes in interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    if intervalT[\"taxon_id\"]==\"4081\":\n",
    "        graph = \"http://solgenomics.net/genome/Solanum_lycopersicum\"\n",
    "    else:\n",
    "        raise Exception(\"unknown taxon_id \"+str(intervalT[\"taxon_id\"]))\n",
    "    genes = requests.get(url+\"/getFeaturesInInterval\", \n",
    "                        params={\"feature\": \"'protein_coding_gene'\", \"graph\": graph,\n",
    "                               \"begin\": intervalT[\"begin_pos\"], \"end\": intervalT[\"end_pos\"],\n",
    "                               \"chrom\": intervalT[\"chrom\"]}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "    genesT_in_interval = []\n",
    "    for gene in genes.json()[\"results\"][\"bindings\"]:\n",
    "        gene_id = gene[\"feature_id\"][\"value\"]\n",
    "        #get location\n",
    "        location = requests.get(url+\"/getFeatureLocation\", \n",
    "                        params={\"featureid\": \"'\"+gene_id+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "        chrom = location.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "        taxon_id = location.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "        begin_pos = int(location.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])\n",
    "        end_pos = int(location.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"])    \n",
    "        genesT_in_interval.append([gene_id, chrom, begin_pos, end_pos, taxon_id])\n",
    "    genesT_in_interval = pd.DataFrame(genesT_in_interval)\n",
    "    genesT_in_interval.columns = [\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"taxon_id\"]\n",
    "    genesT_in_interval = genesT_in_interval.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get genes in interval \"+str(intervalT[\"begin_pos\"])+\" - \"+str(intervalT[\"begin_pos\"])+\" on \"+str(intervalT[\"chrom\"]))    \n",
    "#display(genesT_in_interval)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find potato orthologs for these genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    ortholog_genesT_for_interval = []\n",
    "    for gene_id,gene in genesT_in_interval.iterrows():\n",
    "        orthologs = requests.get(url+\"/getOrthologs\", \n",
    "                        params={\"geneid\": \"'\"+gene_id+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "        orthologsResult = orthologs.json()[\"results\"][\"bindings\"]\n",
    "        if len(orthologsResult)>0:\n",
    "            for ortholog in orthologsResult:\n",
    "                ortholog_id = ortholog[\"ortholog_id\"][\"value\"]\n",
    "                #get location\n",
    "                location = requests.get(url+\"/getFeatureLocation\", \n",
    "                                params={\"featureid\": \"'\"+ortholog_id+\"'\"}, \n",
    "                                headers={\"accept\": \"application/json\"})\n",
    "                chrom = location.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "                taxon_id = location.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "                begin_pos = int(location.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])\n",
    "                end_pos = int(location.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"]) \n",
    "                ortholog_genesT_for_interval.append([gene_id, gene[\"chrom\"],gene[\"begin_pos\"],\n",
    "                                                     gene[\"end_pos\"],gene[\"taxon_id\"],ortholog_id,\n",
    "                                                     chrom, begin_pos, end_pos, taxon_id])\n",
    "        else:\n",
    "            ortholog_genesT_for_interval.append([gene_id, gene[\"chrom\"],gene[\"begin_pos\"],\n",
    "                                                 gene[\"end_pos\"],gene[\"taxon_id\"], None, None,\n",
    "                                                None, None, None])\n",
    "    ortholog_genesT_for_interval = pd.DataFrame(ortholog_genesT_for_interval)            \n",
    "    ortholog_genesT_for_interval.columns = [\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"taxon_id\", \n",
    "                                            \"ortholog_gene_id\", \"ortholog_chrom\", \"ortholog_begin_pos\",\n",
    "                                            \"ortholog_end_pos\", \"ortholog_taxon_id\"]\n",
    "    ortholog_genesT_for_interval = ortholog_genesT_for_interval.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get orthologs\")    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of orthologs for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>ortholog_gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>nunique</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Solyc10g076170.1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076180.1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076190.1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076200.1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076210.1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076220.1</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076230.1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ortholog_gene_id\n",
       "                          nunique\n",
       "gene_id                          \n",
       "Solyc10g076170.1                0\n",
       "Solyc10g076180.1                0\n",
       "Solyc10g076190.1                1\n",
       "Solyc10g076200.1                2\n",
       "Solyc10g076210.1                2\n",
       "Solyc10g076220.1                2\n",
       "Solyc10g076230.1                1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aggregations = { \"ortholog_gene_id\" : [\"nunique\"]}\n",
    "display(ortholog_genesT_for_interval.groupby([\"gene_id\"]).agg(aggregations))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List orthologs\n",
    "List orthologs for the genes found in the interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ortholog_chrom</th>\n",
       "      <th>ortholog_begin_pos</th>\n",
       "      <th>ortholog_end_pos</th>\n",
       "      <th>taxon_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>begin_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>ortholog_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Solyc10g076170.1</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>58981351</th>\n",
       "      <th>58981887</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076180.1</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>59006329</th>\n",
       "      <th>59007294</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076190.1</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>59045991</th>\n",
       "      <th>59047132</th>\n",
       "      <th>PGSC0003DMG400011948</th>\n",
       "      <td>chromosome 0</td>\n",
       "      <td>22096297.0</td>\n",
       "      <td>22097627.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Solyc10g076200.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chromosome 10</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">59051770</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">59052216</th>\n",
       "      <th>PGSC0003DMG400011955</th>\n",
       "      <td>chromosome 0</td>\n",
       "      <td>22098971.0</td>\n",
       "      <td>22100334.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400040954</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>49170543.0</td>\n",
       "      <td>49171657.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Solyc10g076210.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chromosome 10</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">59054478</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">59055612</th>\n",
       "      <th>PGSC0003DMG400020799</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>49151461.0</td>\n",
       "      <td>49152451.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020800</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>49084193.0</td>\n",
       "      <td>49085040.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Solyc10g076220.1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chromosome 10</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">59059828</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">59060961</th>\n",
       "      <th>PGSC0003DMG400020799</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>49151461.0</td>\n",
       "      <td>49152451.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020800</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>49084193.0</td>\n",
       "      <td>49085040.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076230.1</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>59074041</th>\n",
       "      <th>59074647</th>\n",
       "      <th>PGSC0003DMG400020798</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>49172630.0</td>\n",
       "      <td>49173264.0</td>\n",
       "      <td>4081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                       ortholog_chrom  \\\n",
       "gene_id          chrom         begin_pos end_pos  ortholog_gene_id                      \n",
       "Solyc10g076170.1 chromosome 10 58981351  58981887 NaN                            None   \n",
       "Solyc10g076180.1 chromosome 10 59006329  59007294 NaN                            None   \n",
       "Solyc10g076190.1 chromosome 10 59045991  59047132 PGSC0003DMG400011948   chromosome 0   \n",
       "Solyc10g076200.1 chromosome 10 59051770  59052216 PGSC0003DMG400011955   chromosome 0   \n",
       "                                                  PGSC0003DMG400040954  chromosome 10   \n",
       "Solyc10g076210.1 chromosome 10 59054478  59055612 PGSC0003DMG400020799  chromosome 10   \n",
       "                                                  PGSC0003DMG400020800  chromosome 10   \n",
       "Solyc10g076220.1 chromosome 10 59059828  59060961 PGSC0003DMG400020799  chromosome 10   \n",
       "                                                  PGSC0003DMG400020800  chromosome 10   \n",
       "Solyc10g076230.1 chromosome 10 59074041  59074647 PGSC0003DMG400020798  chromosome 10   \n",
       "\n",
       "                                                                        ortholog_begin_pos  \\\n",
       "gene_id          chrom         begin_pos end_pos  ortholog_gene_id                           \n",
       "Solyc10g076170.1 chromosome 10 58981351  58981887 NaN                                  NaN   \n",
       "Solyc10g076180.1 chromosome 10 59006329  59007294 NaN                                  NaN   \n",
       "Solyc10g076190.1 chromosome 10 59045991  59047132 PGSC0003DMG400011948          22096297.0   \n",
       "Solyc10g076200.1 chromosome 10 59051770  59052216 PGSC0003DMG400011955          22098971.0   \n",
       "                                                  PGSC0003DMG400040954          49170543.0   \n",
       "Solyc10g076210.1 chromosome 10 59054478  59055612 PGSC0003DMG400020799          49151461.0   \n",
       "                                                  PGSC0003DMG400020800          49084193.0   \n",
       "Solyc10g076220.1 chromosome 10 59059828  59060961 PGSC0003DMG400020799          49151461.0   \n",
       "                                                  PGSC0003DMG400020800          49084193.0   \n",
       "Solyc10g076230.1 chromosome 10 59074041  59074647 PGSC0003DMG400020798          49172630.0   \n",
       "\n",
       "                                                                        ortholog_end_pos  \\\n",
       "gene_id          chrom         begin_pos end_pos  ortholog_gene_id                         \n",
       "Solyc10g076170.1 chromosome 10 58981351  58981887 NaN                                NaN   \n",
       "Solyc10g076180.1 chromosome 10 59006329  59007294 NaN                                NaN   \n",
       "Solyc10g076190.1 chromosome 10 59045991  59047132 PGSC0003DMG400011948        22097627.0   \n",
       "Solyc10g076200.1 chromosome 10 59051770  59052216 PGSC0003DMG400011955        22100334.0   \n",
       "                                                  PGSC0003DMG400040954        49171657.0   \n",
       "Solyc10g076210.1 chromosome 10 59054478  59055612 PGSC0003DMG400020799        49152451.0   \n",
       "                                                  PGSC0003DMG400020800        49085040.0   \n",
       "Solyc10g076220.1 chromosome 10 59059828  59060961 PGSC0003DMG400020799        49152451.0   \n",
       "                                                  PGSC0003DMG400020800        49085040.0   \n",
       "Solyc10g076230.1 chromosome 10 59074041  59074647 PGSC0003DMG400020798        49173264.0   \n",
       "\n",
       "                                                                       taxon_id  \n",
       "gene_id          chrom         begin_pos end_pos  ortholog_gene_id               \n",
       "Solyc10g076170.1 chromosome 10 58981351  58981887 NaN                      4081  \n",
       "Solyc10g076180.1 chromosome 10 59006329  59007294 NaN                      4081  \n",
       "Solyc10g076190.1 chromosome 10 59045991  59047132 PGSC0003DMG400011948     4081  \n",
       "Solyc10g076200.1 chromosome 10 59051770  59052216 PGSC0003DMG400011955     4081  \n",
       "                                                  PGSC0003DMG400040954     4081  \n",
       "Solyc10g076210.1 chromosome 10 59054478  59055612 PGSC0003DMG400020799     4081  \n",
       "                                                  PGSC0003DMG400020800     4081  \n",
       "Solyc10g076220.1 chromosome 10 59059828  59060961 PGSC0003DMG400020799     4081  \n",
       "                                                  PGSC0003DMG400020800     4081  \n",
       "Solyc10g076230.1 chromosome 10 59074041  59074647 PGSC0003DMG400020798     4081  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "index=pd.MultiIndex.from_tuples([tuple(x) for x in ortholog_genesT_for_interval[[\"chrom\", \"begin_pos\", \"end_pos\", \"ortholog_gene_id\"]].to_records()], names=[\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"ortholog_gene_id\"])\n",
    "display(ortholog_genesT_for_interval.set_index(index)[[\"ortholog_chrom\", \"ortholog_begin_pos\", \"ortholog_end_pos\", \"taxon_id\"]])   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chromosomes for orthologs\n",
    "Chromosomes containing the found orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ortholog_gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>nunique</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ortholog_taxon_id</th>\n",
       "      <th>ortholog_chrom</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">4113</th>\n",
       "      <th>chromosome 0</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromosome 10</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 ortholog_gene_id\n",
       "                                          nunique\n",
       "ortholog_taxon_id ortholog_chrom                 \n",
       "4113              chromosome 0                  2\n",
       "                  chromosome 10                 4"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kauraggregations = {\"ortholog_begin_pos\" : [\"min\",\"max\"], \"ortholog_end_pos\" : [\"min\",\"max\"]}\n",
    "display(ortholog_genesT_for_interval.groupby([\"ortholog_taxon_id\", \"ortholog_chrom\"]).agg(aggregations))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Diagram crosslinks\n",
    "Create a diagram with crosslinks between found genes and orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg fill-rule=\"evenodd\" height=\"595.2755905511812\" preserveAspectRatio=\"xMinYMin meet\" version=\"1.0\" viewBox=\"0 0 841 595\" width=\"841.8897637795277\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "\t<title>...</title>\n",
       "\t<desc>...</desc>\n",
       "\t<clipPath id=\"clip\">\n",
       "\t\t<rect height=\"595.2755905511812\" width=\"841.8897637795277\" x=\"0\" y=\"0\"/>\n",
       "\t</clipPath>\n",
       "\t<g id=\"group\" style=\"clip-path: url(#clip)\" transform=\"scale(1,-1) translate(0,-595)\">\n",
       "\t\t<g transform=\"\">\n",
       "\t\t\t<polygon points=\"42.09449 76.64173, 483.7792 76.64173, 483.7792 197.185, 42.09449 197.185\" style=\"stroke: rgb(96%,96%,96%); stroke-linecap: butt; stroke-width: 1; fill: rgb(96%,96%,96%);\"/>\n",
       "\t\t\t<polygon points=\"42.09449 237.3661, 139.4327 237.3661, 139.4327 357.9094, 42.09449 357.9094\" style=\"stroke: rgb(96%,96%,96%); stroke-linecap: butt; stroke-width: 1; fill: rgb(96%,96%,96%);\"/>\n",
       "\t\t\t<polygon points=\"42.09449 398.0906, 799.7912 398.0906, 799.7912 518.6339, 42.09449 518.6339\" style=\"stroke: rgb(96%,96%,96%); stroke-linecap: butt; stroke-width: 1; fill: rgb(96%,96%,96%);\"/>\n",
       "\t\t\t<polygon points=\"649.9632 398.0906, 654.5837 398.0906, 87.9755 357.9094, 82.58964 357.9094\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"673.3654 398.0906, 675.1714 398.0906, 98.93754 357.9094, 93.41805 357.9094\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"649.9632 398.0906, 654.5837 398.0906, 87.9755 357.9094, 82.58964 357.9094\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"673.3654 398.0906, 675.1714 398.0906, 98.93754 357.9094, 93.41805 357.9094\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"763.5521 398.0906, 766.0061 398.0906, 443.284 197.185, 440.7166 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"684.3314 398.0906, 688.9236 398.0906, 359.0015 197.185, 354.9924 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"684.3314 398.0906, 688.9236 398.0906, 86.01958 197.185, 82.58964 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"673.3654 398.0906, 675.1714 398.0906, 436.7765 197.185, 432.2653 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"673.3654 398.0906, 675.1714 398.0906, 436.7765 197.185, 432.2653 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"684.3314 398.0906, 688.9236 398.0906, 359.0015 197.185, 354.9924 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"705.9963 398.0906, 710.5845 398.0906, 359.0015 197.185, 354.9924 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"763.5521 398.0906, 766.0061 398.0906, 443.284 197.185, 440.7166 197.185\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"440.7166 136.9134, 443.284 136.9134, 443.284 197.185, 440.7166 197.185\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"354.9924 136.9134, 359.0015 136.9134, 359.0015 197.185, 354.9924 197.185\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"82.58964 136.9134, 86.01958 136.9134, 86.01958 197.185, 82.58964 197.185\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"432.2653 136.9134, 436.7765 136.9134, 436.7765 197.185, 432.2653 197.185\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"82.58964 297.6378, 87.9755 297.6378, 87.9755 357.9094, 82.58964 357.9094\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"93.41805 297.6378, 98.93754 297.6378, 98.93754 357.9094, 93.41805 357.9094\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"388.2025 458.3622, 390.3731 458.3622, 390.3731 518.6339, 388.2025 518.6339\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"489.3513 458.3622, 493.2591 458.3622, 493.2591 518.6339, 489.3513 518.6339\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"649.9632 458.3622, 654.5837 458.3622, 654.5837 518.6339, 649.9632 518.6339\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"673.3654 458.3622, 675.1714 458.3622, 675.1714 518.6339, 673.3654 518.6339\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"684.3314 458.3622, 688.9236 458.3622, 688.9236 518.6339, 684.3314 518.6339\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"705.9963 458.3622, 710.5845 458.3622, 710.5845 518.6339, 705.9963 518.6339\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"763.5521 458.3622, 766.0061 458.3622, 766.0061 518.6339, 763.5521 518.6339\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<path d=\"M 483.779188,76.641732 L 483.779188,197.185039 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,136.913386 L 483.779188,136.913386 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,76.641732 L 42.094488,197.185039 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 139.432691,237.366142 L 139.432691,357.909449 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,297.637795 L 139.432691,297.637795 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,237.366142 L 42.094488,357.909449 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 799.791226,398.090551 L 799.791226,518.633858 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,458.362205 L 799.791226,458.362205 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,398.090551 L 42.094488,518.633858 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 463.721938,458.362205 L 463.721938,488.498031 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 58.770393,458.362205 L 58.770393,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 99.265547,458.362205 L 99.265547,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 139.760702,458.362205 L 139.760702,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 180.255857,458.362205 L 180.255857,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 220.751011,458.362205 L 220.751011,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 261.246166,458.362205 L 261.246166,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 301.741320,458.362205 L 301.741320,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 342.236475,458.362205 L 342.236475,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 382.731629,458.362205 L 382.731629,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 423.226784,458.362205 L 423.226784,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 463.721938,458.362205 L 463.721938,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 504.217093,458.362205 L 504.217093,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 544.712248,458.362205 L 544.712248,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 585.207402,458.362205 L 585.207402,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 625.702557,458.362205 L 625.702557,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 666.197711,458.362205 L 666.197711,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 706.692866,458.362205 L 706.692866,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 747.188020,458.362205 L 747.188020,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 787.683175,458.362205 L 787.683175,476.443701 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,463.721938,488.498031)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">59000000</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,440.716641,197.185039)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020798</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,354.992448,197.185039)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020799</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,82.589643,197.185039)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020800</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,432.265302,197.185039)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400040954</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,82.589643,357.909449)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400011948</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,93.418047,357.909449)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400011955</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,388.202525,518.633858)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076170.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,489.351322,518.633858)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076180.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,649.963204,518.633858)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076190.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,673.365354,518.633858)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076200.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,684.331441,518.633858)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076210.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,705.996349,518.633858)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076220.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,763.552112,518.633858)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076230.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,42.000000,76.641732)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,193.000000,76.641732)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,344.000000,76.641732)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,42.000000,237.366142)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,42.000000,398.090551)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,193.000000,398.090551)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,344.000000,398.090551)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,495.000000,398.090551)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,646.000000,398.090551)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t</g>\n",
       "\t</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genesTdiagram = DIAGRAMS.crosslinks(\"tomato\", \"potato\", intervalT, ortholog_genesT_for_interval)\n",
    "tmpFilename = \"tomato_\"+tg1+\"-\"+tg2+\"_crosslinks.svg\"\n",
    "genesTdiagram.write(tmpFilename, \"svg\")\n",
    "SVG(tmpFilename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Potato\n",
    "### Find interval\n",
    "Find locations for genes, and compute interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chrom': 'chromosome 10',\n",
       " 'taxon_id': '4113',\n",
       " 'begin_pos': 48982521,\n",
       " 'end_pos': 49258672}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    resp1 = requests.get(url+\"/getFeatureLocation\", \n",
    "                        params={\"featureid\": \"'\"+pg1+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "    resp2 = requests.get(url+\"/getFeatureLocation\", \n",
    "                        params={\"featureid\": \"'\"+pg2+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "    intervalP = {}\n",
    "    intervalP[\"chrom\"] = resp1.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "    intervalP[\"taxon_id\"] = resp1.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "    intervalP[\"begin_pos\"] = int(resp1.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"])\n",
    "    intervalP[\"end_pos\"] = int(resp2.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])    \n",
    "except:\n",
    "    raise Exception(\"couldn't get interval for \"+str(pg1)+\" or \"+str(pg2))\n",
    "intervalP   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find genes and orthologs\n",
    "Find genes and orthologs for interval"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find potato genes in interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    if intervalP[\"taxon_id\"]==\"4113\":\n",
    "        graph = \"http://solgenomics.net/genome/Solanum_tuberosum\"\n",
    "    else:\n",
    "        raise Exception(\"unknown taxon_id \"+str(intervalP[\"taxon_id\"]))\n",
    "    genes = requests.get(url+\"/getFeaturesInInterval\", \n",
    "                        params={\"feature\": \"'protein_coding_gene'\", \"graph\": graph,\n",
    "                               \"begin\": intervalP[\"begin_pos\"], \"end\": intervalP[\"end_pos\"],\n",
    "                               \"chrom\": intervalP[\"chrom\"]}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "    genesP_in_interval = []\n",
    "    for gene in genes.json()[\"results\"][\"bindings\"]:\n",
    "        gene_id = gene[\"feature_id\"][\"value\"]\n",
    "        #get location\n",
    "        location = requests.get(url+\"/getFeatureLocation\", \n",
    "                        params={\"featureid\": \"'\"+gene_id+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "        chrom = location.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "        taxon_id = location.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "        begin_pos = int(location.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])\n",
    "        end_pos = int(location.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"])    \n",
    "        genesP_in_interval.append([gene_id, chrom, begin_pos, end_pos, taxon_id])\n",
    "    genesP_in_interval = pd.DataFrame(genesP_in_interval)\n",
    "    genesP_in_interval.columns = [\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"taxon_id\"]\n",
    "    genesP_in_interval = genesP_in_interval.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get genes in interval \"+str(intervalP[\"begin_pos\"])+\" - \"+str(intervalP[\"begin_pos\"])+\" on \"+str(intervalP[\"chrom\"]))    \n",
    "#display(genesP_in_interval)    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find tomato orthologs for these genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    ortholog_genesP_for_interval = []\n",
    "    for gene_id,gene in genesP_in_interval.iterrows():\n",
    "        orthologs = requests.get(url+\"/getOrthologs\", \n",
    "                        params={\"geneid\": \"'\"+gene_id+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "        orthologsResult = orthologs.json()[\"results\"][\"bindings\"]\n",
    "        if len(orthologsResult)>0:\n",
    "            for ortholog in orthologsResult:\n",
    "                ortholog_id = ortholog[\"ortholog_id\"][\"value\"]\n",
    "                #get location\n",
    "                location = requests.get(url+\"/getFeatureLocation\", \n",
    "                                params={\"featureid\": \"'\"+ortholog_id+\"'\"}, \n",
    "                                headers={\"accept\": \"application/json\"})\n",
    "                chrom = location.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "                taxon_id = location.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "                begin_pos = int(location.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])\n",
    "                end_pos = int(location.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"]) \n",
    "                ortholog_genesP_for_interval.append([gene_id, gene[\"chrom\"],gene[\"begin_pos\"],\n",
    "                                                     gene[\"end_pos\"],gene[\"taxon_id\"],ortholog_id,\n",
    "                                                     chrom, begin_pos, end_pos, taxon_id])\n",
    "        else:\n",
    "            ortholog_genesP_for_interval.append([gene_id, gene[\"chrom\"],gene[\"begin_pos\"],\n",
    "                                                 gene[\"end_pos\"],gene[\"taxon_id\"], None, None,\n",
    "                                                None, None, None])\n",
    "    ortholog_genesP_for_interval = pd.DataFrame(ortholog_genesP_for_interval)            \n",
    "    ortholog_genesP_for_interval.columns = [\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"taxon_id\", \n",
    "                                            \"ortholog_gene_id\", \"ortholog_chrom\", \"ortholog_begin_pos\",\n",
    "                                            \"ortholog_end_pos\", \"ortholog_taxon_id\"]\n",
    "    ortholog_genesP_for_interval = ortholog_genesP_for_interval.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get orthologs\")  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of orthologs for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>ortholog_gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>nunique</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400006679</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400006680</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400006681</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020795</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020797</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020798</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020799</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020800</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400035649</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400039458</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400040544</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400040954</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400045482</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ortholog_gene_id\n",
       "                              nunique\n",
       "gene_id                              \n",
       "PGSC0003DMG400006679                0\n",
       "PGSC0003DMG400006680                0\n",
       "PGSC0003DMG400006681                0\n",
       "PGSC0003DMG400020795                0\n",
       "PGSC0003DMG400020797                0\n",
       "PGSC0003DMG400020798                2\n",
       "PGSC0003DMG400020799                2\n",
       "PGSC0003DMG400020800                2\n",
       "PGSC0003DMG400035649                0\n",
       "PGSC0003DMG400039458                0\n",
       "PGSC0003DMG400040544                2\n",
       "PGSC0003DMG400040954                1\n",
       "PGSC0003DMG400045482                0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "aggregations = { \"ortholog_gene_id\" : [\"nunique\"]}\n",
    "display(ortholog_genesP_for_interval.groupby([\"gene_id\"]).agg(aggregations))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### List orthologs\n",
    "List orthologs for the genes found in the interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ortholog_chrom</th>\n",
       "      <th>ortholog_begin_pos</th>\n",
       "      <th>ortholog_end_pos</th>\n",
       "      <th>taxon_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>begin_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>ortholog_gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400006679</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49022687</th>\n",
       "      <th>49023709</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400006680</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49040786</th>\n",
       "      <th>49046010</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400006681</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49064472</th>\n",
       "      <th>49065184</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020795</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49247297</th>\n",
       "      <th>49254377</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020797</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49199138</th>\n",
       "      <th>49201970</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">PGSC0003DMG400020798</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chromosome 10</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49172630</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49173264</th>\n",
       "      <th>Solyc10g075120.1</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>58847992.0</td>\n",
       "      <td>58848834.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076230.1</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>59074041.0</td>\n",
       "      <td>59074647.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">PGSC0003DMG400020799</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chromosome 10</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49151461</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49152451</th>\n",
       "      <th>Solyc10g076210.1</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>59054478.0</td>\n",
       "      <td>59055612.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076220.1</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>59059828.0</td>\n",
       "      <td>59060961.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">PGSC0003DMG400020800</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chromosome 10</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49084193</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49085040</th>\n",
       "      <th>Solyc10g076210.1</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>59054478.0</td>\n",
       "      <td>59055612.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076220.1</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>59059828.0</td>\n",
       "      <td>59060961.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400035649</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49174387</th>\n",
       "      <th>49174710</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400039458</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49235999</th>\n",
       "      <th>49240574</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">PGSC0003DMG400040544</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chromosome 10</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49108641</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">49111957</th>\n",
       "      <th>Solyc01g066890.1</th>\n",
       "      <td>chromosome 1</td>\n",
       "      <td>75142075.0</td>\n",
       "      <td>75142941.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc07g041360.1</th>\n",
       "      <td>chromosome 7</td>\n",
       "      <td>52534390.0</td>\n",
       "      <td>52534818.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400040954</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49170543</th>\n",
       "      <th>49171657</th>\n",
       "      <th>Solyc10g076200.1</th>\n",
       "      <td>chromosome 10</td>\n",
       "      <td>59051770.0</td>\n",
       "      <td>59052216.0</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400045482</th>\n",
       "      <th>chromosome 10</th>\n",
       "      <th>49092329</th>\n",
       "      <th>49093552</th>\n",
       "      <th>NaN</th>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                       ortholog_chrom  \\\n",
       "gene_id              chrom         begin_pos end_pos  ortholog_gene_id                  \n",
       "PGSC0003DMG400006679 chromosome 10 49022687  49023709 NaN                        None   \n",
       "PGSC0003DMG400006680 chromosome 10 49040786  49046010 NaN                        None   \n",
       "PGSC0003DMG400006681 chromosome 10 49064472  49065184 NaN                        None   \n",
       "PGSC0003DMG400020795 chromosome 10 49247297  49254377 NaN                        None   \n",
       "PGSC0003DMG400020797 chromosome 10 49199138  49201970 NaN                        None   \n",
       "PGSC0003DMG400020798 chromosome 10 49172630  49173264 Solyc10g075120.1  chromosome 10   \n",
       "                                                      Solyc10g076230.1  chromosome 10   \n",
       "PGSC0003DMG400020799 chromosome 10 49151461  49152451 Solyc10g076210.1  chromosome 10   \n",
       "                                                      Solyc10g076220.1  chromosome 10   \n",
       "PGSC0003DMG400020800 chromosome 10 49084193  49085040 Solyc10g076210.1  chromosome 10   \n",
       "                                                      Solyc10g076220.1  chromosome 10   \n",
       "PGSC0003DMG400035649 chromosome 10 49174387  49174710 NaN                        None   \n",
       "PGSC0003DMG400039458 chromosome 10 49235999  49240574 NaN                        None   \n",
       "PGSC0003DMG400040544 chromosome 10 49108641  49111957 Solyc01g066890.1   chromosome 1   \n",
       "                                                      Solyc07g041360.1   chromosome 7   \n",
       "PGSC0003DMG400040954 chromosome 10 49170543  49171657 Solyc10g076200.1  chromosome 10   \n",
       "PGSC0003DMG400045482 chromosome 10 49092329  49093552 NaN                        None   \n",
       "\n",
       "                                                                        ortholog_begin_pos  \\\n",
       "gene_id              chrom         begin_pos end_pos  ortholog_gene_id                       \n",
       "PGSC0003DMG400006679 chromosome 10 49022687  49023709 NaN                              NaN   \n",
       "PGSC0003DMG400006680 chromosome 10 49040786  49046010 NaN                              NaN   \n",
       "PGSC0003DMG400006681 chromosome 10 49064472  49065184 NaN                              NaN   \n",
       "PGSC0003DMG400020795 chromosome 10 49247297  49254377 NaN                              NaN   \n",
       "PGSC0003DMG400020797 chromosome 10 49199138  49201970 NaN                              NaN   \n",
       "PGSC0003DMG400020798 chromosome 10 49172630  49173264 Solyc10g075120.1          58847992.0   \n",
       "                                                      Solyc10g076230.1          59074041.0   \n",
       "PGSC0003DMG400020799 chromosome 10 49151461  49152451 Solyc10g076210.1          59054478.0   \n",
       "                                                      Solyc10g076220.1          59059828.0   \n",
       "PGSC0003DMG400020800 chromosome 10 49084193  49085040 Solyc10g076210.1          59054478.0   \n",
       "                                                      Solyc10g076220.1          59059828.0   \n",
       "PGSC0003DMG400035649 chromosome 10 49174387  49174710 NaN                              NaN   \n",
       "PGSC0003DMG400039458 chromosome 10 49235999  49240574 NaN                              NaN   \n",
       "PGSC0003DMG400040544 chromosome 10 49108641  49111957 Solyc01g066890.1          75142075.0   \n",
       "                                                      Solyc07g041360.1          52534390.0   \n",
       "PGSC0003DMG400040954 chromosome 10 49170543  49171657 Solyc10g076200.1          59051770.0   \n",
       "PGSC0003DMG400045482 chromosome 10 49092329  49093552 NaN                              NaN   \n",
       "\n",
       "                                                                        ortholog_end_pos  \\\n",
       "gene_id              chrom         begin_pos end_pos  ortholog_gene_id                     \n",
       "PGSC0003DMG400006679 chromosome 10 49022687  49023709 NaN                            NaN   \n",
       "PGSC0003DMG400006680 chromosome 10 49040786  49046010 NaN                            NaN   \n",
       "PGSC0003DMG400006681 chromosome 10 49064472  49065184 NaN                            NaN   \n",
       "PGSC0003DMG400020795 chromosome 10 49247297  49254377 NaN                            NaN   \n",
       "PGSC0003DMG400020797 chromosome 10 49199138  49201970 NaN                            NaN   \n",
       "PGSC0003DMG400020798 chromosome 10 49172630  49173264 Solyc10g075120.1        58848834.0   \n",
       "                                                      Solyc10g076230.1        59074647.0   \n",
       "PGSC0003DMG400020799 chromosome 10 49151461  49152451 Solyc10g076210.1        59055612.0   \n",
       "                                                      Solyc10g076220.1        59060961.0   \n",
       "PGSC0003DMG400020800 chromosome 10 49084193  49085040 Solyc10g076210.1        59055612.0   \n",
       "                                                      Solyc10g076220.1        59060961.0   \n",
       "PGSC0003DMG400035649 chromosome 10 49174387  49174710 NaN                            NaN   \n",
       "PGSC0003DMG400039458 chromosome 10 49235999  49240574 NaN                            NaN   \n",
       "PGSC0003DMG400040544 chromosome 10 49108641  49111957 Solyc01g066890.1        75142941.0   \n",
       "                                                      Solyc07g041360.1        52534818.0   \n",
       "PGSC0003DMG400040954 chromosome 10 49170543  49171657 Solyc10g076200.1        59052216.0   \n",
       "PGSC0003DMG400045482 chromosome 10 49092329  49093552 NaN                            NaN   \n",
       "\n",
       "                                                                       taxon_id  \n",
       "gene_id              chrom         begin_pos end_pos  ortholog_gene_id           \n",
       "PGSC0003DMG400006679 chromosome 10 49022687  49023709 NaN                  4113  \n",
       "PGSC0003DMG400006680 chromosome 10 49040786  49046010 NaN                  4113  \n",
       "PGSC0003DMG400006681 chromosome 10 49064472  49065184 NaN                  4113  \n",
       "PGSC0003DMG400020795 chromosome 10 49247297  49254377 NaN                  4113  \n",
       "PGSC0003DMG400020797 chromosome 10 49199138  49201970 NaN                  4113  \n",
       "PGSC0003DMG400020798 chromosome 10 49172630  49173264 Solyc10g075120.1     4113  \n",
       "                                                      Solyc10g076230.1     4113  \n",
       "PGSC0003DMG400020799 chromosome 10 49151461  49152451 Solyc10g076210.1     4113  \n",
       "                                                      Solyc10g076220.1     4113  \n",
       "PGSC0003DMG400020800 chromosome 10 49084193  49085040 Solyc10g076210.1     4113  \n",
       "                                                      Solyc10g076220.1     4113  \n",
       "PGSC0003DMG400035649 chromosome 10 49174387  49174710 NaN                  4113  \n",
       "PGSC0003DMG400039458 chromosome 10 49235999  49240574 NaN                  4113  \n",
       "PGSC0003DMG400040544 chromosome 10 49108641  49111957 Solyc01g066890.1     4113  \n",
       "                                                      Solyc07g041360.1     4113  \n",
       "PGSC0003DMG400040954 chromosome 10 49170543  49171657 Solyc10g076200.1     4113  \n",
       "PGSC0003DMG400045482 chromosome 10 49092329  49093552 NaN                  4113  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "index=pd.MultiIndex.from_tuples([tuple(x) for x in ortholog_genesP_for_interval[[\"chrom\", \"begin_pos\", \"end_pos\", \"ortholog_gene_id\"]].to_records()], names=[\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"ortholog_gene_id\"])\n",
    "display(ortholog_genesP_for_interval.set_index(index)[[\"ortholog_chrom\", \"ortholog_begin_pos\", \"ortholog_end_pos\", \"taxon_id\"]])   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Chromosomes for orthologs\n",
    "Chromosomes containing the found orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ortholog_gene_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>nunique</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ortholog_taxon_id</th>\n",
       "      <th>ortholog_chrom</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">4081</th>\n",
       "      <th>chromosome 1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromosome 10</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromosome 7</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 ortholog_gene_id\n",
       "                                          nunique\n",
       "ortholog_taxon_id ortholog_chrom                 \n",
       "4081              chromosome 1                  1\n",
       "                  chromosome 10                 5\n",
       "                  chromosome 7                  1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "kauraggregations = {\"ortholog_begin_pos\" : [\"min\",\"max\"], \"ortholog_end_pos\" : [\"min\",\"max\"]}\n",
    "display(ortholog_genesP_for_interval.groupby([\"ortholog_taxon_id\", \"ortholog_chrom\"]).agg(aggregations))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Diagram crosslinks\n",
    "Create a diagram with crosslinks between found genes and orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<svg fill-rule=\"evenodd\" height=\"595.2755905511812\" preserveAspectRatio=\"xMinYMin meet\" version=\"1.0\" viewBox=\"0 0 841 595\" width=\"841.8897637795277\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "\t<title>...</title>\n",
       "\t<desc>...</desc>\n",
       "\t<clipPath id=\"clip\">\n",
       "\t\t<rect height=\"595.2755905511812\" width=\"841.8897637795277\" x=\"0\" y=\"0\"/>\n",
       "\t</clipPath>\n",
       "\t<g id=\"group\" style=\"clip-path: url(#clip)\" transform=\"scale(1,-1) translate(0,-595)\">\n",
       "\t\t<g transform=\"\">\n",
       "\t\t\t<polygon points=\"42.09449 71.61909, 98.14446 71.61909, 98.14446 162.0266, 42.09449 162.0266\" style=\"stroke: rgb(96%,96%,96%); stroke-linecap: butt; stroke-width: 1; fill: rgb(96%,96%,96%);\"/>\n",
       "\t\t\t<polygon points=\"42.09449 192.1624, 718.8619 192.1624, 718.8619 282.5699, 42.09449 282.5699\" style=\"stroke: rgb(96%,96%,96%); stroke-linecap: butt; stroke-width: 1; fill: rgb(96%,96%,96%);\"/>\n",
       "\t\t\t<polygon points=\"42.09449 312.7057, 99.34624 312.7057, 99.34624 403.1132, 42.09449 403.1132\" style=\"stroke: rgb(96%,96%,96%); stroke-linecap: butt; stroke-width: 1; fill: rgb(96%,96%,96%);\"/>\n",
       "\t\t\t<polygon points=\"42.09449 433.249, 799.7925 433.249, 799.7925 523.6565, 42.09449 523.6565\" style=\"stroke: rgb(96%,96%,96%); stroke-linecap: butt; stroke-width: 1; fill: rgb(96%,96%,96%);\"/>\n",
       "\t\t\t<polygon points=\"388.1402 433.249, 397.2386 433.249, 71.90842 403.1132, 69.5323 403.1132\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"388.1402 433.249, 397.2386 433.249, 71.90842 403.1132, 69.5323 403.1132\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"563.7121 433.249, 565.4516 433.249, 71.84257 282.5699, 69.5323 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"557.9858 433.249, 561.0424 433.249, 629.8784 282.5699, 628.6546 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"505.629 433.249, 508.3453 433.249, 639.1962 282.5699, 636.0848 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"505.629 433.249, 508.3453 433.249, 653.8727 282.5699, 650.764 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"563.7121 433.249, 565.4516 433.249, 691.4241 282.5699, 689.7614 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"563.7121 433.249, 565.4516 433.249, 71.84257 282.5699, 69.5323 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"505.629 433.249, 508.3453 433.249, 639.1962 282.5699, 636.0848 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"321.0603 433.249, 323.3842 433.249, 639.1962 282.5699, 636.0848 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"557.9858 433.249, 561.0424 433.249, 629.8784 282.5699, 628.6546 282.5699\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"388.1402 433.249, 397.2386 433.249, 70.70664 162.0266, 69.5323 162.0266\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"388.1402 433.249, 397.2386 433.249, 70.70664 162.0266, 69.5323 162.0266\" style=\"stroke: rgb(50%,50%,50%); stroke-linecap: round; stroke-width: 1; fill: rgb(82%,82%,82%);\"/>\n",
       "\t\t\t<polygon points=\"69.5323 116.8228, 70.70664 116.8228, 70.70664 162.0266, 69.5323 162.0266\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"69.5323 237.3661, 71.84257 237.3661, 71.84257 282.5699, 69.5323 282.5699\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"628.6546 237.3661, 629.8784 237.3661, 629.8784 282.5699, 628.6546 282.5699\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"636.0848 237.3661, 639.1962 237.3661, 639.1962 282.5699, 636.0848 282.5699\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"650.764 237.3661, 653.8727 237.3661, 653.8727 282.5699, 650.764 282.5699\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"689.7614 237.3661, 691.4241 237.3661, 691.4241 282.5699, 689.7614 282.5699\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"69.5323 357.9094, 71.90842 357.9094, 71.90842 403.1132, 69.5323 403.1132\" style=\"stroke: rgb(100%,0%,0%); stroke-linecap: butt; stroke-width: 1; fill: rgb(100%,0%,0%);\"/>\n",
       "\t\t\t<polygon points=\"152.3012 478.4528, 155.1054 478.4528, 155.1054 523.6565, 152.3012 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"201.9609 478.4528, 216.2944 478.4528, 216.2944 523.6565, 201.9609 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"266.9501 478.4528, 268.9037 478.4528, 268.9037 523.6565, 266.9501 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"768.582 478.4528, 788.008 478.4528, 788.008 523.6565, 768.582 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"636.4442 478.4528, 644.2146 478.4528, 644.2146 523.6565, 636.4442 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"563.7121 478.4528, 565.4516 478.4528, 565.4516 523.6565, 563.7121 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"505.629 478.4528, 508.3453 478.4528, 508.3453 523.6565, 505.629 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"321.0603 478.4528, 323.3842 478.4528, 323.3842 523.6565, 321.0603 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"568.5329 478.4528, 569.4191 478.4528, 569.4191 523.6565, 568.5329 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"737.5828 478.4528, 750.1356 478.4528, 750.1356 523.6565, 737.5828 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"388.1402 478.4528, 397.2386 478.4528, 397.2386 523.6565, 388.1402 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"557.9858 478.4528, 561.0424 478.4528, 561.0424 523.6565, 557.9858 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<polygon points=\"343.3837 478.4528, 346.7393 478.4528, 346.7393 523.6565, 343.3837 523.6565\" style=\"stroke: rgb(0%,0%,100%); stroke-linecap: butt; stroke-width: 1; fill: rgb(0%,0%,100%);\"/>\n",
       "\t\t\t<path d=\"M 98.144460,71.619094 L 98.144460,162.026575 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,116.822835 L 98.144460,116.822835 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,71.619094 L 42.094488,162.026575 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 718.861949,192.162402 L 718.861949,282.569882 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,237.366142 L 718.861949,237.366142 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,192.162402 L 42.094488,282.569882 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 99.346236,312.705709 L 99.346236,403.113189 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,357.909449 L 99.346236,357.909449 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,312.705709 L 42.094488,403.113189 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 799.792532,433.249016 L 799.792532,523.656496 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,478.452756 L 799.792532,478.452756 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 42.094488,433.249016 L 42.094488,523.656496 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 90.053047,478.452756 L 90.053047,501.054626 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 62.615231,478.452756 L 62.615231,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 90.053047,478.452756 L 90.053047,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 117.490864,478.452756 L 117.490864,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 144.928680,478.452756 L 144.928680,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 172.366497,478.452756 L 172.366497,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 199.804313,478.452756 L 199.804313,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 227.242129,478.452756 L 227.242129,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 254.679946,478.452756 L 254.679946,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 282.117762,478.452756 L 282.117762,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 309.555579,478.452756 L 309.555579,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 336.993395,478.452756 L 336.993395,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 364.431211,478.452756 L 364.431211,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 391.869028,478.452756 L 391.869028,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 419.306844,478.452756 L 419.306844,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 446.744661,478.452756 L 446.744661,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 474.182477,478.452756 L 474.182477,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 501.620293,478.452756 L 501.620293,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 529.058110,478.452756 L 529.058110,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 556.495926,478.452756 L 556.495926,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 583.933743,478.452756 L 583.933743,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 611.371559,478.452756 L 611.371559,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 638.809375,478.452756 L 638.809375,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 666.247192,478.452756 L 666.247192,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 693.685008,478.452756 L 693.685008,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 721.122825,478.452756 L 721.122825,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 748.560641,478.452756 L 748.560641,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<path d=\"M 775.998457,478.452756 L 775.998457,492.013878 Z\" style=\"stroke: rgb(0%,0%,0%); stroke-linecap: butt; stroke-width: 1;\"/>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,90.053047,501.054626)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">49000000</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,69.532305,162.026575)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc07g041360.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,69.532305,282.569882)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g075120.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,628.654640,282.569882)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076200.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,636.084800,282.569882)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076210.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,650.764032,282.569882)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076220.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,689.761400,282.569882)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc10g076230.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,69.532305,403.113189)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">Solyc01g066890.1</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,152.301222,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400006679</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,201.960925,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400006680</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,266.950137,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400006681</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,768.582016,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020795</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,636.444236,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020797</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,563.712072,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020798</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,505.628958,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020799</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,321.060255,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400020800</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,568.532896,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400035649</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,737.582771,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400039458</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,388.140229,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400040544</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,557.985800,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400040954</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(0.707107,0.707107,-0.707107,0.707107,343.383662,523.656496)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 6px; fill: rgb(0%,0%,0%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">PGSC0003DMG400045482</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,42.000000,192.162402)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,193.000000,192.162402)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,344.000000,192.162402)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,495.000000,192.162402)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,646.000000,192.162402)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">tomato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,42.000000,433.249016)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,193.000000,433.249016)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,344.000000,433.249016)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,495.000000,433.249016)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t\t<g transform=\" matrix(1.000000,0.000000,-0.000000,1.000000,646.000000,433.249016)\">\n",
       "\t\t\t\t<text style=\"font-family: Helvetica; font-size: 8px; fill: rgb(60%,60%,60%);\" transform=\"translate(0,0) scale(1,-1)\" x=\"0\" y=\"0\">potato</text>\n",
       "\t\t\t</g>\n",
       "\t\t</g>\n",
       "\t</g>\n",
       "</svg>"
      ],
      "text/plain": [
       "<IPython.core.display.SVG object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genesPdiagram = DIAGRAMS.crosslinks(\"potato\", \"tomato\", intervalP, ortholog_genesP_for_interval)\n",
    "tmpFilename = \"potato_\"+pg1+\"-\"+pg2+\".svg\"\n",
    "genesPdiagram.write(tmpFilename, \"svg\")\n",
    "SVG(tmpFilename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotations\n",
    "Compare GO annotatios in all three classes\n",
    "\n",
    "Genes, orthologs for potato and tomato and annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "setT = set(genesT_in_interval.index.unique())\n",
    "setP = set(genesP_in_interval.index.unique())\n",
    "setTP = setT & set(ortholog_genesP_for_interval[\"ortholog_gene_id\"].dropna().unique())\n",
    "setPT = setP & set(ortholog_genesT_for_interval[\"ortholog_gene_id\"].dropna().unique())\n",
    "setT = setT - setTP\n",
    "setP = setP - setPT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "setT(3) = {'Solyc10g076190.1', 'Solyc10g076170.1', 'Solyc10g076180.1'}\n",
      "setP(9) = {'PGSC0003DMG400006681', 'PGSC0003DMG400020795', 'PGSC0003DMG400045482', 'PGSC0003DMG400039458', 'PGSC0003DMG400006680', 'PGSC0003DMG400035649', 'PGSC0003DMG400020797', 'PGSC0003DMG400006679', 'PGSC0003DMG400040544'}\n",
      "setTP(4) = {'Solyc10g076200.1', 'Solyc10g076220.1', 'Solyc10g076210.1', 'Solyc10g076230.1'}\n"
     ]
    }
   ],
   "source": [
    "# debug\n",
    "print('setT(%d) = %s' % (len(setT), setT))\n",
    "print('setP(%d) = %s' % (len(setP), setP))\n",
    "print('setTP(%d) = %s' % (len(setTP), setTP))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQsAAADrCAYAAAB+ZrNkAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAHy1JREFUeJztnXlwXNd1p7/TC7obAEGABLhAgEhREiWukijKoizJEmU5llRJ7JlKxZOkUi577Bq7Zso1nqmMU7LsNrzJSxKnUtFkapKxy1OxNY7s2CUnNh1JdLRQKyVSpijuKwhwA0iQAHvvvvPH645ACiQbYPe773afr+oVGg32u6eb7/36nHPPPVeMMSiKolyOkG0DFEVxAxULRVGqQsVCUZSqULFQFKUqVCwURakKFQtFUapCxUJRlKpQsVAUpSpULBRFqQoVC0VRqkLFQlGUqlCxUBSlKlQsFEWpChULRVGqQsVCUZSqULFQFKUqVCwURakKFQtFUapCxUJRlKpQsVAUpSpULBRFqQoVC0VRqkLFQlGUqojYNkBxBxmQCJAAYkB80hGb4nEEKEw68hf8nPx8CjgDnDVJU/DvHSnTQXSTIWUqZEBiQPcFRwcgdR66IhxjwClgFDhlkiZX53GVy6BioSADkuDdwjDLqlHvZhwYAY4AgyZpJizb03SoWLiESAvvuPsRoFg+SpMen//7FP/BMiACzAcWlY9OP8yvMaeBQeAwcMwkTcmyPQ2PikVQkH/7dp8LtHN+TqByTDchXcJz6yfyUcbfWsvZf1hJ61Oz6T0Qp+1UpGFyVnlgCE881OuoEyoWfiMiwGw8UZjLOwKRqPVQBszRfjKD12JG5xEvhc8Xm6xQPNJCYWeC4putyLZWWibChGtthwVOA7uAXSZpsraNaRRULPxApBvoB/qAHuo8C5VJUNh/A9nBJcTysemNNRIhtztOcVsrbGklPBSjpV52+kAB2Au8bZJmxLYxrqNiUQ9EInjisKj8s+Zew1ScWEhm342eF0GoNrMWIxFyL7aT3zibyL44sVqc0xIngO3AfpM0RdvGuIiKRa3wko+Ly0cfPtWwlARzcCnp/TcSybTW1ws4HSb/Sju5jR1EdrQ6KxwZYCeet6G5jWmgYnGliMwBVgLX4WORW0kwh64nvWcF0VycqF/jVjgTJv/sLHI/mUvM0USpAQ4Cr5mkGbNsixOoWMwEkRCeB7ESWODn0CXBHLqO9N4VRLMJ/0XiQopg3mwl8+O5yLZW4rbtmQElYAfwukmajG1jgoyKxXTwpjeXlY82v4c/vITUrtXBEImpGI6SfbKL4r/MJp4PObfuKAe8AbylNRtTo2JRDSKdwBpgCRYW352dTW7rHZTOdrnxzZ0Sihs6yTzeTSLjnmicBV4xSXPAtiFBQ8XiUojEgbXAjVgQiUKY0ttrSB9eQmutZjf85FyIwhNzyP1sDomiOGf/MeAlkzQnbRsSFFQspsKb+lwF3AR26gyGFpHevoaIjeRlrTkdJv/9HvLPzKbVti0zYA+eaDR9PkPF4kJEluJ5E+02hs/GKWy+i/zpHn9qM/xkOEr2b+dhNre7EU5NIgU8a5Jm0LYhNlGxqCDSC6zDK7+2wvFe0lvuIFpocXIqsmq2JUh/q5fomHtTrm8DLzdrzw0VC5EocCew1JYJJcFsu43U4LX+z7DYIiMU/24e2V91OheajAEbm7F8vLnFQmQBsB6LvRsmZpF79R5MapazFZFXxNtx0t/uJToSdcrLKOLlMd62bYifNKdYeEVVtwI3U//OTxfl8BJS29YSN2HnphdrSlYofr+H7M+7nPMy9gLPm6TJ2zbED5pPLLyaifV4qz+t8dYaUgdvcO7mqCs746QH+pxbJj8GPGWS5rRtQ+pNc4mFyHK8JKY1l7cYovTaPWRGFqhQTMVYmPyX+ig5tsI1B2wwSXPMtiH1pDnEwktirsdbz2GNbJzCS/dRnJjt1I3gO3ko/c18Mk+5lfwsAE+bpDls25B60fhiIdIOPADMsWnGmU5yr6xHGqHIyi+e6uDcYwtodaj6swT8q0mavbYNqQeNLRYi84AP4lPzmYtxqpvsy+uJlCJOxeKBYF+MzBf7iJ5167PbZJJmu20jak3jioXINcB9YPciG+0h88p6Wi7sf6lUz5kw+c/1Yxxr8bfZJM0bto2oJY0pFl4i804sTosCjMwj8+q9KhS1YCJE4U+vpnTILcF4yyTNi7aNqBWNJxYit+LVUFhFhaL2pITiI/0U9iScShDvBp5rhB4ZjXUhi9yJCkXD0moIPzpIZFUKl1aALgXusW1ELWici1nkNmCFbTPG5pBVoagfMUN4YJCWWyecEozrZUDW2DbiSmmMC1pkNXCLbTPSCQovryesQlFfohB6ZIiWO8ZJ27ZlGqyVAbnOthFXgvsXtdd/Yp1tMwoRSi/dT7HRl5cHhQiE/mSY2Aq3QpJ7ZEDm2zZiprgtFiKLgPfZNsOAeeVesql2pxJvzhOF0JeOEF2UJWfblioJAx+UAQnaDvVV4a5YiCwE7icA72HLHaQbsbOVC8QN4a8NIt15XGlIEwcelAFxaQoYCMCNNiO8vUM/iOWCK4A9y0kNL3ZqDUPDMbtI9GuDFGMlXJme7AQ+IAPi1P3nlLEAiLQBD2Kpke5kRnvI7FqlHkUQ6M0TSx4hKwZXCoeuAu6ybcR0cEssRASvhNv6DZprofj6XYRdbNHfqKxKk/jUCVK27ZgGN8qALLNtRLW4JRZewdVC20YAbL6bnK4gDR4PjdF2u1tTqutcSXi6IxYiVxGAWgqAfTeSOjXPvnejTM1njxHtLDiT8IziSIWnG2Lh7TG6HssLwwDGO8jtXO3cvhdNRVuJyBeGcKkvZq8MyHLbRlyO4IvFO3kK6zMOBsxr78M0e4NdF1iaIfGHI07lL24PejjiwkV/C17m2Dp7VpBq1pb9LvL7oyRuSJO1bUeVBD4cCbZYePt6WF9FCt66j70rNE/hEmGQzw8RirtTfxHocCS4YuHt7XE3AchTALy5jpwuEHOPriLRTx13av1IYMORIF/8K4Au20YAHLuKtLbud5f1Z0lck3Fm/Uhgw5FgioU3+xGI8KMYorTtNl1J6jIhkP921JlQBLxw5HrbRlxIMMUCbicA5dwAO28inU1o8ZXrLM4R/8CYU7Mja4O2diRQxgCV9v3WdjSfTDZG8dD1mtRsFD5+khaHkp2zgEAlO4MlFl5NxZ22zaiwazUZTWo2Du0lIp884VQp+C0yIIHxaoN2I9yA5Q2LK2TjFAavUa+i0Xj/GVr73WmWkwBW2zaiQnDEQqQFeI9tMyrsuImsVmo2HmGQj590Zt0IwKqgNMoJ0s2wEoKx5iKToDC0WL2KRmXNORILc86sHWnBuzesEwyxEIkQkA8EYMfN5EwoIJ+NUnNCIB896YxYAKyUAbE+fR+UG2IZAfEqci0Uh/uDYYtSP9ZNkOjJOyMYcQIwM2JfLLyy7sAkcQ4sJaO5isYnDPLHI86IBcBq23UXQbgprgXabBtR4fB1wSgGU+rP3WeJO9QkpxVYZNOAIIjFKtsGVDh2FRmt1mweIhD6g1FnlrCD5WJFu2LhLUHvtmrDJPbf6ExnaKVG3HuGeNidjuD9MiDW8mm2PYvAzICkWsmf6tHEZrPRagi/d9yZJewhwNp+qfbEQiQOLLY2/gXsX0YOCUbvDMVfHhyzbcG0sBaK2PQsrrE8/nkc7dNcRbOyPE28o0DRth1V0i0DYqXPi82bdYnFsc/jTCe5bKvOgjQrYZAHx5wJRQCs9LqwIxZeCBKIzYIABpc4Nd+u1IEPnHHKs7xeBsT3kNmWZ7HY4tjv4piGIE3P/AIt16edWY3aBvT6PaitGzYwIch4B7lMm4YgCtx31ikP0/dQxH+xEIlhQRUvxuASZyr4lDpzyzmneq32+T2gDc9isaVxp+REL2HbNijBYGGeFodmRVplQDr8HNDGTRuYEKQQoTTRoSGI4hECuX3CqfJvXycJ/BULb4VpYEKQ0XlktRBLmcx7ztm2YFos8HMwvz2LbgiO239ygTNrAhSfWJlyytNsYM8C5vk83iUZnRcc4VKCQXuJiEO7l3XIgPi2U17TikUxRGlittZXKO9m3YRTU6i+eRdNKxanu7XPpjI1SzNO5bF8y1v4d7N4Jd6+TvVcitF5zkyRKT7Tn3Wq3qIhPYvAeBUA451OfXsoPtJdIOpQQ5wuGZCYHwM1rVhMdGgIokxNGGSxO7uWCTDHj4GaVixSbZrcVC7O0rRTYWq7H4P4KRa+qF81ZBIUShGdNlUuznUu1XF6O67XHX/EwqvcDMx2gOOzdfGYcmmuyTgVpjaUZ9EKwUkonu2kZNsGJdgsyDs1I9JQYhGYTYQAJjqcyXQrlmgtORWmqljUi1zMKRdTsUAYxKHl6ioW9SKru4MoVdBVdEYsIn5sPtSUYpFvUc9CuTxzCk7lturuXahYKMpF6M6rWEymKcWiEHUqeaVYYk7BqUR4w4hFYGosiiFKpbB6FsrlmeNKxsKj7lO9fs0lX9HNOQKRVfAnBYiUIHwbvL4Bfj6Tc+VjlK7UnkbjW5u4b+sx7gbkpgU8/7k7eca2TUGgrViH2qACwnf4PDHG+Ax/XcMz1/2a9uumuaJxuqDwKvzFSfjKQfjKb2DF//T2Sp02JXHKtaw7zx2id+sx7n7sIR793of48u5RVr86FKx1PLaI1ONK+UfeTxtH63BmFQvwmnb243VdPgfhIoRnekITmDrSYLB7lIXz29nflSAXi1C6eja7n9rPzbbtCgKhWlcdH6KTIVaxmhdqel4PFYsKGZB58IV++LMVsONTcGBGJ1KxOI/lPQwdHWfp4TO0nU7Tsv80K8cywVn0Z5Oa97T4OR/hHn5Cfbzbut/LztS/x8GcgK/sgcT98On/B73/AYZt2+U67+3n2OZhNjz8DP81Gibb3cqRkDg1ZegGG1lFnHHWcJhNLLVtzkzwSyxqdvFdD+kVsPsJWDEjsdCMxbv4zO1sAjYBPLKRD3clOG3ZpEBQrOWeMoNcx3Fu4uusxBClSJy/4eN8mu/WaIS6C7wTYvEmtLdC8XpIn4DoNlj2Udgwk3OJisW72HeaWdd2Mb7tOHP2neaWb6/lm7ZtCgKlWn61fJSfAj8FYBNLeZ3fqqFQQAOJxRV96G/B7M/CxwyEDMjtsPmrsG0m5woZzVpcyFee5VPZIm0hofjvl/F4Xwcp2zYFgYJbV0rdxUKM8eGrVuQjwOz6D3R5CmFKG35f6yyUy7NhNuceWxCs6uNL8KpJmq31HMCvmybt0ziXJVIkJEVN4CmX55Qz6X+A+nd/80ssAuXWRvPOLD1WLDIScSpkrfuWzn6JRaD2po7mdE5EuTynIk6FqxP1HqBZxULDEOWyOCYW4/UeoCnFoiWrnoVyeUbd2S6iYJImU+9BmlUs1LNQLkkJzLg7rQx8ub+aUizazzqVuFIskApRNLWs4Kwvdc9XQJOKRceYM98YiiVORJ3aiKru+QrwSyyMKRGgWotZY+4soFPscCDmVKjaUJ4FEJzFSYk0kZA7e0IoFtjj1nYRDScWJ3wc67K0nnPKzVR8Zm/cKe9TxaKetI2rZ6FMTQnM/phTYnHKj0GaVixmndFaC2VqTkXI50POJMHH/KixAD/FwpgUPrlL1TD3hDMFN4rPHG5xKkStR/PfKfFbPQPjXXSdpIWSehfKu9kbd+q6ULGoN5EiofZxr2O4okzm5Xan8hXH/BrIb7E47vN4l2TOCafm0hUfSAnFPQlitu2okgmTNL6F9n6LxQg+tP+qlp5jzpTzKj7xdqtT3qZvIQj4LRbGFPH5DV6K7uO0UOu9IRSnedWVJnoevoUgYGfPz5ltDlQHonnCrRPkbNuhBIdXZtFi24Zp0MCehccBArR7x7xhp6bJlDpyPELuVMSZ5GbGJM2YnwP6LxbGpAlQKNJ/gKhtG5RgsKWNvG0bpsGQ3wPaqlILTCgy+zQtsbRTF4lSJzZ2OONVAOzxe0CbYhGYUGTBEc1bNDsjEXI7Wp2ZMk0DR/we1I5YeKXfvmZyL0XfAae+UZQ6sLHDKe9yr0ka30sQbC6WCUwo0jVKrCXj1MWi1JASmF904lIHC99DELArFvsJUCgyf0hDkWZld5zMaNSZhYWnTNKM2BjYnlh4ochha+NfwLU7dFakWfllp20LpoUVrwLsehYAb1ke/99oH6elczQ4fUIVf8gKxednOZPYNDStWBgzRIB6c16z07YFit+8MIusQ41uhkzSWNs3OAgfUmC8i97DxDXR2TwUwfyg26nwc7fNwYMgFnsIyDYBAtK/XxOdzcLL7aRPRp0Riwxw0KYB9sXCmAKwzbYZFZbsIqYdtBqfEpjv9zgjFABvmaSxuo7Jvlh4vA3B+EaPZYgsGAqGp6PUjzfaSB9tcUYscgQgXA+GWBiTwxOMQLBsC1H1Lhqb/9vtVNXudpM01r9MgyEWHr8hIN5F2zmiCwfVu2hUtiVIH4g707ciMGF6cMTCmAzwum0zKizbSlRKwWkBqNSGEpjv9QTour882/3aF+RyBO1D205A6i5aU0R7DxOI/ySldmyaRdqhhrw5YKttIyoESyy83dY32TajwrKttEhRvYtGISWU/tc8pxaMbTVJE5gGwsESCwBjhvEWmVknnibSf0BzF43CD7vJnI0E8JqfmhQBmAGZTFA/uJchGL0xl28hEc0GwxZl5gxHyT7ZRcK2HdPgddt1FRcSTLEwZoKAxGqRAqEVb2gJuOv85QIw4sw+MSeAwK1UCqZYeLwJjNs2AqDvIInOEQ1HXGVTOymHWuYVgX81SRO4Op/gioW3IdHzts2ocMtLRDTZ6R7jIQp/vcAZoQDY7HeL/2oJrlgAGHMEz8OwTtsE0Wt261SqS5TAfLuXwkTYmS5YJ/CKEwNJsMXC4zUCsqHyjW+SiKeCUWWqXJ5/7iS9pc2ZqdLAhh8Vgi8WXu3FM2B/w9qQQdY+j9F1I8HncAuZ/zPPqdmPwIYfFYIvFlCZHXnWthkAnaeIXbdDk51BJisUv9RHuOjW7Edgw48KbogFgDEHCUiRyg2/IdFxSvMXQeWx+WQdamoT+PCjgjti4fEyYKUN+mQE5D3PEYnktFgraDw/i9SvZ9Nq245pEPjwo4JbYuHlL54mAEvZ42kia16kgNH8RVDYFyPz5wudylPsN0kTiNm+anBLLACMOQv8igCUg887SnzJTqx1W1beYSRC7uF+oo7lKX5t24jp4J5YABhzFG+GxHqR1PKttPUMq2DYJCUUH+5HUu7UU4wDvzJJU7RtyHRwUywAjDlEQGZIbnuOxKwxTXjaICeUvthPwbF+mhtM0jg3o+auWAAYswd4ybYZIYOs20g0pgVbvlIE841esrvcaWZTAp42SROIBk/TxW2xADBmG7DFthmxLOE7NiLhPE65lq5SAvPYfNKvtTuV0HzRJM0R20bMFPfFAsCY1whAd/D2caK3PUdee3fWlyKYv1hI5qlOp6ZIt5mksX6NXgmNIRYem4C9to3oPkF8zSayKhj1IQ+lR3vJPNvhlEdxCK9GyGnEBL9wrHpEBHgvsMK2KUf7SL9xJzHjzqa7gScnlL56FTmHFoeBJxRPuzbzMRWNJRYVRNYAa22bcewq0q/fpYJRC7JC8Qt9FBxqYgPePr7PmqRpCC+zMcUCQORG4C4sh1rHe0lvvlsF40pICcWHr6awL+6UULxlkuZF20bUksYVCwCRPuB+sLv7lArGzBmOkn2kn5BDC8PAW+/xhm0jak1jiwWASCfwANBh04zRHjKv3kO0GHWmytA6r7aR+lYv8axbIrvJJM1220bUg8YXCwCROPABYKFNM861kX/p/ZhMmzP7bFqhCOYH3aSfmOvU1GgJb6m59Rm5etEcYgGVmZKbgVuxmMfIRym+sp7c2Fynpv5841yIwqO9FN50a8ajgDfjcdi2IfWkecSigkgPsB7otGVCSTBb15EeXuzUN2fdGWwh+8U+wiNRIrZtmQYTeEJxwrYh9ab5xAJAJAKsA5bbNGPPclK7VpEg5Myy6rpQgNKP55L+4VxaHdoICGAQ+HVQdjmvN80pFhVErgbuAXshwem5ZDffTSibcCrbXzMOt5D5Ri+hwZhTeRyDt71gw814XIrmFguoJD/vARbZMqEQobTlDjLH+5onLMlD6fFuMj+eQ8Ixb+IcnjcxbNsQv1GxqCCyBHgPFqdYDy8h9datxEqRxp5e3Rcj881ewg71oKiwD3jBJI31bSlsoGIxGZEQ3rqSNWCnWvBcG/nN76M43unUbEBVnA6T/34P+WfcaqgLXsOaFxp5WrQaVCymQqQFTzBWgJ1v+QNLSe28iVixAbyMcyEKP5pL7skuEg71yKxwGE8oJmwbYhsVi0shMgsvNLnWxvDZGMVta8keu9q5b2LAWyX6T52kf9hNwrEqTIBTwMsuN6upNSoW1SAyD6+Yq9/G8KM9ZN5ch6Ta3VhIlYfScx1kvttD7Kx7nlEa2AzsdGHjHz9RsZgOIrPxQpOl+Lw4zYDZu5z0vmW0FFqCWbQ0Fib/i05yP+8i7tDO5RWKwDZgi0mavG1jgoiKxUzwirqW4hV1zfFz6EKY0r5lpPffSDwoi9L2xMj8ZC7mxXbijk2DVtgLvKp5iUujYnGliPTieRuL8HHNSSFMae8KMgeWErMhGlmh+HI72SfmEjnkVkHVZI4CrzRDqXYtULGoFSJtwBLgGmA++PMNW4hQ2ruc9MHridU7PEkJxS1tZH/dgWxuJ+7gzAZ4i772AG+bpBm1bYxLqFjUA5EEsBjP2+iF+ucYSoIZXkR6/w2Ezs6pXY3GRIjC623kNnYQ2tJGzNEwA2AMrwP8bpM0ur/LDFCxqDciYbw+Gv1AH9BV7yHHO8jtW05huH/61aDnQhT2x8hvb6W0tZXw2wmnBcLgNczdbpJmyLYxrqNi4TciUbykaDcwt3zMoQ7FX4UwpcElZIauQca6iF+4urUE5mSE/O4EhW0J2NpG1MES7KlIAbuAHZq0rB0qFkHAa8zTiScc3UAbEMdbDRsvHzNNnuaBc+kEZ7bdxvgvlpJ4bhbzD8aIHWkhnHevWGoqSni7kh8GjpikGbFsT0OiYuEKXgl6fNIRxqsNKJV/Fqf4PY95d3wuAxLBC4muLh8uVoim8PpJDOIJhOYh6oyKhYIMSBeeR1M55mK5I/oU5IAR4AgwqDMZ/qNioUyJDEgH5wtIN/iyErYAjOPNXozirdEYNUkz7sPYyiVQsVCqRgakjfPzKHG8pfxTPY7g3fj5C34Wpng+BZwFzpikOeffO1Kmg4qFoihV0QiZcEVRfEDFQlGUqlCxUBSlKlQsFEWpChULRVGqQsVCUZSqULGYhIhMTDpKIpKe9Psf+WhHXESMiPT5NaaiXI5A9nK0hTGmvfJYRA4CnzDGPG3PIkUJDupZTAMRSYjIYyJyVESOiMi3xVtyjog8ICJ7ReQRERkRkSEReUhEPiQi+0RkVET++6Rz3Skir4jIGREZFpHviNfbE+C58s9dZa/mw+XX/OdJ5/pHEZnv7yegNDMqFtNjAFgNrMLbGuBe4H9M+vsivPLlBcA3gO8Cv1d+zf3A10TkqvK/zQP/Ba+Xxd3A7wCfKP/tfeWfNxhj2o0xPxORh4AvAP8OuApvUdXf1/4tKspFMMboMcUBHATuv+C5IeC+Sb9/CNhZfvwAcAYIlX/vwevUdNOkf78deOAi4/0p8Hj5cbz82r5Jf/8B8OVJv3fiLUdfYPuzqvKzTAMTwHHge0D7ZV7zJeDvpzHGvcAR2++1kQ/1LKpEvAY1C/DatFU4hPctX+GkMaZUfpwu/zw+6e9poL18vuUi8ksROS4iZ4Ev4q3svBi9k8c2xozhLb666qKvCBa/Y7yc0BrgNuARy/Yo00TFokqM9/V1DC/UqHA1nrcxE/4WeAO41hjTAXyZdzqCT7W6b3jy2OJteNRxBeNbwRgzBPwSWCkivSLypIicKud7Pgle/gd4GPhIOWfzZvn5j4nIDhEZF5H9IvKfys+3lc/ZO2n2qldEYiLyl+Wc0HD5sRO7ugURFYvp8TiQFJG54m1p+HlmnjeYBZwxxkyIyArgk5U/GGOyeCHNkgvG/qSIrBSROPBNYKMx5tgMx7eCiPQDDwFb8N7TETyv6feAr4vI+40xG4CvAz8yXs7mpvLLTwC/jSeSHwO+IyJrjDHngAeB4fK/bzfGDOP9/6wDbgZuwtu3Vj2aGaJiMT2+iNdOfjuwFdgEfGuG5/os8AkRmQAeA340xVhPiMiYiPyuMeafgEeBJ/G8jAXAH89wbBv8TETGgBeAZ4H/DdwFfM4YkzHGbAX+jku8J2PMPxtj9hmPZ4F/wUsOX4w/wsvznDDGnMRLULv0mQUKrbO4CMaYxVM8lwI+XT4u/NsG4LpJv09wwUZDxpi1kx4/g7cF4sXG/yvgry73nEN82EyqWRGR24FTxpzXAesQsPZdr3znNQ8CSbzPLYTXO3TbJcY8L89Tftw7fdMVUM9CsccwMEdEZk16bnIO6Ly8TTnX8BPgz4D5xphO4BdMI89TPv/wlZvenKhYKFYwxgwCLwKPlsvbVwP/EW+KGLxZpMUiUrlGW/Da9p0ECmUv47cmnfI4MLec+K3wOPCIiPSISDdeaKe1KTNExUKxyR/gbfM4DPwUSBpjnir/7Ynyz1EReaMcrnwG+AfgNPCHePkbAIwxO/HEYX85z9MLfBXYDPwGL1x5o/ycMgO0B6eiKFWhnoWiKFWhYqEoSlWoWCiKUhUqFoqiVIWKhaIoVaFioShKVahYKIpSFSoWiqJUxf8HccJV/FkfROsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "venn2(subsets = (len(setT), len(setTP), len(setP)), set_labels = (\"Tomato\", \"Potato\"))\n",
    "#venn2(subsets = (1,2,3), set_labels = (\"Tomato\", \"Potato\"))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Only in Tomato interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solyc10g076190.1\n",
      "Solyc10g076170.1\n",
      "Solyc10g076180.1\n"
     ]
    }
   ],
   "source": [
    "print(*setT, sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    termsT = []\n",
    "    for gene_id in setT:\n",
    "        annotations = requests.get(url+\"/getGeneAnnotations\", \n",
    "                                   params={\"geneid\": \"'\"+gene_id+\"'\"}, \n",
    "                                   headers={\"accept\": \"application/json\"})        \n",
    "        for annotation in annotations.json()[\"results\"][\"bindings\"]:\n",
    "            uniprot_goa = annotation[\"uniprot_goa\"][\"value\"].strip()\n",
    "            if uniprot_goa==\"\":\n",
    "                termsT.append([gene_id, None])\n",
    "            else:\n",
    "                for go_id in uniprot_goa.split(\",\"):                    \n",
    "                    termsT.append([gene_id, go_id.strip()])\n",
    "            break        \n",
    "    termsT = pd.DataFrame(termsT)\n",
    "    termsT.columns = [\"gene_id\", \"go_id\"]\n",
    "    termsT = termsT.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get terms for genes\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO:0004601 [peroxidase activity]; GO:0005576 [extracellular region]; GO:0006979 [response to oxidative stress]; GO:0020037 [heme binding]; GO:0042744 [hydrogen peroxide catabolic process]; GO:0046872 [metal ion binding]\n",
      "DNA-templated]\n",
      "GO:0003677 [DNA binding]; GO:0045892 [negative regulation of transcription\n"
     ]
    }
   ],
   "source": [
    "print(*set([go_id for go_id in termsT[\"go_id\"] if not (go_id == None)]), sep=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th>go_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Solyc10g076190.1</th>\n",
       "      <th>GO:0004601 [peroxidase activity]; GO:0005576 [extracellular region]; GO:0006979 [response to oxidative stress]; GO:0020037 [heme binding]; GO:0042744 [hydrogen peroxide catabolic process]; GO:0046872 [metal ion binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Solyc10g076170.1</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Solyc10g076180.1</th>\n",
       "      <th>GO:0003677 [DNA binding]; GO:0045892 [negative regulation of transcription</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA-templated]</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [(Solyc10g076190.1, GO:0004601 [peroxidase activity]; GO:0005576 [extracellular region]; GO:0006979 [response to oxidative stress]; GO:0020037 [heme binding]; GO:0042744 [hydrogen peroxide catabolic process]; GO:0046872 [metal ion binding]), (Solyc10g076170.1, nan), (Solyc10g076180.1, GO:0003677 [DNA binding]; GO:0045892 [negative regulation of transcription), (Solyc10g076180.1, DNA-templated])]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "termsT.set_index([\"go_id\"], append=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Only in Potato"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PGSC0003DMG400006681\n",
      "PGSC0003DMG400020795\n",
      "PGSC0003DMG400045482\n",
      "PGSC0003DMG400039458\n",
      "PGSC0003DMG400006680\n",
      "PGSC0003DMG400035649\n",
      "PGSC0003DMG400020797\n",
      "PGSC0003DMG400006679\n",
      "PGSC0003DMG400040544\n"
     ]
    }
   ],
   "source": [
    "print(*setP, sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    termsP = []\n",
    "    for gene_id in setP:\n",
    "        annotations = requests.get(url+\"/getGeneAnnotations\", \n",
    "                                   params={\"geneid\": \"'\"+gene_id+\"'\"}, \n",
    "                                   headers={\"accept\": \"application/json\"})        \n",
    "        for annotation in annotations.json()[\"results\"][\"bindings\"]:\n",
    "            uniprot_goa = annotation[\"uniprot_goa\"][\"value\"].strip()\n",
    "            if uniprot_goa==\"\":\n",
    "                termsP.append([gene_id, None])\n",
    "            else:\n",
    "                for go_id in uniprot_goa.split(\";\"):                    \n",
    "                    termsP.append([gene_id, go_id.strip()])\n",
    "            break        \n",
    "    termsP = pd.DataFrame(termsP)\n",
    "    termsP.columns = [\"gene_id\", \"go_id\"]\n",
    "    termsP = termsP.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get terms for genes\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO:0006979 [response to oxidative stress]\n",
      "GO:0042744 [hydrogen peroxide catabolic process]\n",
      "GO:0046872 [metal ion binding]\n",
      "GO:0004601 [peroxidase activity]\n",
      "GO:0020037 [heme binding]\n"
     ]
    }
   ],
   "source": [
    "print(*set([go_id for go_id in termsP[\"go_id\"] if not (go_id == None)]), sep=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th>go_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">PGSC0003DMG400006681</th>\n",
       "      <th>GO:0004601 [peroxidase activity]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006979 [response to oxidative stress]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0020037 [heme binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">PGSC0003DMG400020795</th>\n",
       "      <th>GO:0004601 [peroxidase activity]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006979 [response to oxidative stress]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0020037 [heme binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400045482</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400039458</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">PGSC0003DMG400006680</th>\n",
       "      <th>GO:0004601 [peroxidase activity]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006979 [response to oxidative stress]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0020037 [heme binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400035649</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400020797</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">PGSC0003DMG400006679</th>\n",
       "      <th>GO:0004601 [peroxidase activity]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006979 [response to oxidative stress]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0020037 [heme binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042744 [hydrogen peroxide catabolic process]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0046872 [metal ion binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGSC0003DMG400040544</th>\n",
       "      <th>NaN</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [(PGSC0003DMG400006681, GO:0004601 [peroxidase activity]), (PGSC0003DMG400006681, GO:0006979 [response to oxidative stress]), (PGSC0003DMG400006681, GO:0020037 [heme binding]), (PGSC0003DMG400020795, GO:0004601 [peroxidase activity]), (PGSC0003DMG400020795, GO:0006979 [response to oxidative stress]), (PGSC0003DMG400020795, GO:0020037 [heme binding]), (PGSC0003DMG400045482, nan), (PGSC0003DMG400039458, nan), (PGSC0003DMG400006680, GO:0004601 [peroxidase activity]), (PGSC0003DMG400006680, GO:0006979 [response to oxidative stress]), (PGSC0003DMG400006680, GO:0020037 [heme binding]), (PGSC0003DMG400035649, nan), (PGSC0003DMG400020797, nan), (PGSC0003DMG400006679, GO:0004601 [peroxidase activity]), (PGSC0003DMG400006679, GO:0006979 [response to oxidative stress]), (PGSC0003DMG400006679, GO:0020037 [heme binding]), (PGSC0003DMG400006679, GO:0042744 [hydrogen peroxide catabolic process]), (PGSC0003DMG400006679, GO:0046872 [metal ion binding]), (PGSC0003DMG400040544, nan)]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "termsP.set_index([\"go_id\"], append=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Both in Tomato and Potato"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solyc10g076200.1\n",
      "Solyc10g076220.1\n",
      "Solyc10g076210.1\n",
      "Solyc10g076230.1\n"
     ]
    }
   ],
   "source": [
    "print(*setTP, sep = \"\\n\")\n",
    "#print(*setPT, sep = \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    termsTP = []\n",
    "    for gene_id in setTP:\n",
    "        annotations = requests.get(url+\"/getGeneAnnotations\", \n",
    "                                   params={\"geneid\": \"'\"+gene_id+\"'\"}, \n",
    "                                   headers={\"accept\": \"application/json\"})        \n",
    "        for annotation in annotations.json()[\"results\"][\"bindings\"]:\n",
    "            uniprot_goa = annotation[\"uniprot_goa\"][\"value\"].strip()\n",
    "            if uniprot_goa==\"\":\n",
    "                termsTP.append([gene_id, None])\n",
    "            else:\n",
    "                for go_id in uniprot_goa.split(\";\"):                    \n",
    "                    termsTP.append([gene_id, go_id.strip()])\n",
    "            break        \n",
    "    termsTP = pd.DataFrame(termsTP)\n",
    "    termsTP.columns = [\"gene_id\", \"go_id\"]\n",
    "    termsTP = termsTP.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get terms for genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GO:0006869 [lipid transport]\n",
      "GO:0016020 [membrane]\n",
      "GO:0006979 [response to oxidative stress]\n",
      "GO:0042744 [hydrogen peroxide catabolic process]\n",
      "GO:0008289 [lipid binding]\n",
      "GO:0004601 [peroxidase activity]\n",
      "GO:0020037 [heme binding]\n",
      "GO:0046872 [metal ion binding]\n",
      "GO:0005576 [extracellular region]\n"
     ]
    }
   ],
   "source": [
    "print(*set([go_id for go_id in termsTP[\"go_id\"] if not (go_id == None)]), sep=\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th>go_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Solyc10g076200.1</th>\n",
       "      <th>GO:0006869 [lipid transport]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0008289 [lipid binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0016020 [membrane]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Solyc10g076220.1</th>\n",
       "      <th>GO:0004601 [peroxidase activity]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005576 [extracellular region]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006979 [response to oxidative stress]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0020037 [heme binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042744 [hydrogen peroxide catabolic process]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0046872 [metal ion binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Solyc10g076210.1</th>\n",
       "      <th>GO:0004601 [peroxidase activity]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0005576 [extracellular region]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006979 [response to oxidative stress]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0020037 [heme binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0042744 [hydrogen peroxide catabolic process]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0046872 [metal ion binding]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Solyc10g076230.1</th>\n",
       "      <th>GO:0004601 [peroxidase activity]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0006979 [response to oxidative stress]</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GO:0020037 [heme binding]</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [(Solyc10g076200.1, GO:0006869 [lipid transport]), (Solyc10g076200.1, GO:0008289 [lipid binding]), (Solyc10g076200.1, GO:0016020 [membrane]), (Solyc10g076220.1, GO:0004601 [peroxidase activity]), (Solyc10g076220.1, GO:0005576 [extracellular region]), (Solyc10g076220.1, GO:0006979 [response to oxidative stress]), (Solyc10g076220.1, GO:0020037 [heme binding]), (Solyc10g076220.1, GO:0042744 [hydrogen peroxide catabolic process]), (Solyc10g076220.1, GO:0046872 [metal ion binding]), (Solyc10g076210.1, GO:0004601 [peroxidase activity]), (Solyc10g076210.1, GO:0005576 [extracellular region]), (Solyc10g076210.1, GO:0006979 [response to oxidative stress]), (Solyc10g076210.1, GO:0020037 [heme binding]), (Solyc10g076210.1, GO:0042744 [hydrogen peroxide catabolic process]), (Solyc10g076210.1, GO:0046872 [metal ion binding]), (Solyc10g076230.1, GO:0004601 [peroxidase activity]), (Solyc10g076230.1, GO:0006979 [response to oxidative stress]), (Solyc10g076230.1, GO:0020037 [heme binding])]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "termsTP.set_index([\"go_id\"], append=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploring the Annotations of Gene *Solyc10g076180.1*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- GO Annotation\n",
    "- PPI \n",
    "- STRING\n",
    "- KEGG\n",
    "- ALL Orthologs\n",
    "- Species with no Orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    ortholog_paralog_genesT_for_interval = []\n",
    "    for gene_id,gene in genesT_in_interval.iterrows():\n",
    "        #get paralogs, and then orthologs\n",
    "        paralogs = requests.get(url+\"/getParalogs\", \n",
    "                        params={\"geneid\": \"'\"+gene_id+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "        for paralog in paralogs.json()[\"results\"][\"bindings\"]:\n",
    "            paralog_id = paralog[\"paralog_id\"][\"value\"]\n",
    "            #get orthologs for paralog\n",
    "            orthologs = requests.get(url+\"/getOrthologs\", \n",
    "                        params={\"geneid\": \"'\"+paralog_id+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "            for ortholog in orthologs.json()[\"results\"][\"bindings\"]:\n",
    "                ortholog_id = ortholog[\"ortholog_id\"][\"value\"]\n",
    "                #get location\n",
    "                location = requests.get(url+\"/getFeatureLocation\", \n",
    "                                params={\"featureid\": \"'\"+ortholog_id+\"'\"}, \n",
    "                                headers={\"accept\": \"application/json\"})\n",
    "                chrom = location.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "                taxon_id = location.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "                begin_pos = int(location.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])\n",
    "                end_pos = int(location.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"]) \n",
    "                ortholog_paralog_genesT_for_interval.append([gene_id, gene[\"chrom\"],gene[\"begin_pos\"],\n",
    "                                                             gene[\"end_pos\"],gene[\"taxon_id\"],paralog_id,\n",
    "                                                             \"PARALOG\", ortholog_id,\n",
    "                                                             chrom, begin_pos, end_pos, taxon_id])\n",
    "        #get direct orthologs\n",
    "        orthologs = requests.get(url+\"/getOrthologs\", \n",
    "                        params={\"geneid\": \"'\"+gene_id+\"'\"}, \n",
    "                        headers={\"accept\": \"application/json\"})\n",
    "        for ortholog in orthologs.json()[\"results\"][\"bindings\"]:\n",
    "            ortholog_id = ortholog[\"ortholog_id\"][\"value\"]\n",
    "            #get location\n",
    "            location = requests.get(url+\"/getFeatureLocation\", \n",
    "                            params={\"featureid\": \"'\"+ortholog_id+\"'\"}, \n",
    "                            headers={\"accept\": \"application/json\"})\n",
    "            chrom = location.json()[\"results\"][\"bindings\"][0][\"chrom\"][\"value\"]\n",
    "            taxon_id = location.json()[\"results\"][\"bindings\"][0][\"taxon_id\"][\"value\"]\n",
    "            begin_pos = int(location.json()[\"results\"][\"bindings\"][0][\"begin_pos\"][\"value\"])\n",
    "            end_pos = int(location.json()[\"results\"][\"bindings\"][0][\"end_pos\"][\"value\"]) \n",
    "            ortholog_paralog_genesT_for_interval.append([gene_id, gene[\"chrom\"],gene[\"begin_pos\"],\n",
    "                                                 gene[\"end_pos\"],gene[\"taxon_id\"],None, \"ORTHOLOG\", ortholog_id,\n",
    "                                                 chrom, begin_pos, end_pos, taxon_id])    \n",
    "    #create dataframe        \n",
    "    ortholog_paralog_genesT_for_interval = pd.DataFrame(ortholog_paralog_genesT_for_interval)            \n",
    "    ortholog_paralog_genesT_for_interval.columns = [\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"taxon_id\", \"paralog_gene_id\",\n",
    "                                            \"path\", \"ortholog_gene_id\", \"ortholog_chrom\", \"ortholog_begin_pos\",\n",
    "                                            \"ortholog_end_pos\", \"ortholog_taxon_id\"]\n",
    "    ortholog_paralog_genesT_for_interval = ortholog_paralog_genesT_for_interval.set_index([\"gene_id\"])\n",
    "except:\n",
    "    raise Exception(\"couldn't get orthologs\")  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of orthologs for each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aggregations = { \"ortholog_gene_id\" : [\"nunique\"]}\n",
    "display(ortholog_paralog_genesT_for_interval.groupby([\"gene_id\"]).agg(aggregations))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "index=pd.MultiIndex.from_tuples([tuple(x) for x in ortholog_paralog_genesT_for_interval[[\"chrom\", \"begin_pos\", \"end_pos\", \"paralog_gene_id\", \"ortholog_gene_id\"]].to_records()], names=[\"gene_id\", \"chrom\", \"begin_pos\", \"end_pos\", \"paralog_gene_id\", \"ortholog_gene_id\"])\n",
    "display(ortholog_paralog_genesT_for_interval.set_index(index)[[\"path\", \"ortholog_chrom\", \"ortholog_begin_pos\", \"ortholog_end_pos\"]])      "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
